dali/pipeline/operators/detection/box_encoder.cc

```diff
@@ -0,0 +1,159 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+
+#include "dali/pipeline/operators/detection/box_encoder.h"
+
+namespace dali {
+namespace detail {
+using Anchor = BoxEncoder<CPUBackend>::Anchor;
+using BBox = BoxEncoder<CPUBackend>::BBox;
+
+float Iou(const Anchor a1, const Anchor a2) {
+  float l = std::max(a1.x, a2.x);
+  float t = std::max(a1.y, a2.y);
+  float r = std::min(a1.z, a2.z);
+  float b = std::min(a1.w, a2.w);
+
+  float width = r - l;
+  float height = b - t;
+
+  if (width <= 0.f || height <= 0.f) return 0.f;
+
+  float intersection = width * height;
+  float area1 = (a1.w - a1.y) * (a1.z - a1.x);
+  float area2 = (a2.w - a2.y) * (a2.z - a2.x);
+
+  return intersection / (area1 + area2 - intersection);
+}
+
+vector<float> CalculateIous(const Anchor *anchors, const BBox *bboxes, int N, int M) {
+  vector<float> ious(N * M);
+
+  for (int bbox_idx = 0; bbox_idx < N; ++bbox_idx) {
+    int best_idx = -1;
+    float best_iou = -1.;
+
+    for (int anchor_idx = 0; anchor_idx < M; ++anchor_idx) {
+      ious[bbox_idx * M + anchor_idx] = detail::Iou(bboxes[bbox_idx], anchors[anchor_idx]);
+
+      if (ious[bbox_idx * M + anchor_idx] >= best_iou) {
+        best_iou = ious[bbox_idx * M + anchor_idx];
+        best_idx = anchor_idx;
+      }
+    }
+
+    // For best default box matched with current object let iou = 2, to make sure there is a match,
+    // as this object will be the best (highest IOU), for this default box
+    ious[bbox_idx * M + best_idx] = 2.;
+  }
+  return ious;
+}
+
+Anchor ToCenterWH(const Anchor a) {
+  return {
+    0.5f * (a.x + a.z),
+    0.5f * (a.y + a.w),
+    a.z - a.x,
+    a.w - a.y};
+}
+}  // namespace detail
+
+template <>
+void BoxEncoder<CPUBackend>::FindMatchingAnchors(
+  const float *ious, const int64 N, const BBox *bboxes,
+  const int *labels, BBox *out_boxes, int *out_labels) {
+  // For every anchor we are looking, if there is a match
+  for (int anchor_idx = 0; anchor_idx < M_; ++anchor_idx) {
+    int best_idx = -1;
+    float best_iou = -1.;
+
+    // Looking for bbox with best IOU with current anchor
+    for (int bbox_idx = 0; bbox_idx < N; ++bbox_idx) {
+      if (ious[bbox_idx * M_ + anchor_idx] >= best_iou) {
+        best_iou = ious[bbox_idx * M_ + anchor_idx];
+        best_idx = bbox_idx;
+      }
+    }
+
+    out_labels[anchor_idx] = 0;
+
+    // Filter matches by criteria
+    // We only report a match, when IOU > criteria
+    if (best_iou > criteria_) {
+      out_boxes[anchor_idx] = bboxes[best_idx];
+      out_labels[anchor_idx] = labels[best_idx];
```

Comment:
What is the value of `out_labels` where there is no match?

Reply:
It is 0, for background. It is set in [here](https://github.com/NVIDIA/DALI/blob/e77f65a158a3cfc269a248aec030a7da755312bd/dali/pipeline/operators/detection/box_encoder.cc#L61)